coalib/misc/Caching.py

```diff
@@ -0,0 +1,131 @@
+import calendar
+import time
+import hashlib
+
+from coalib.misc.CachingUtilities import (
+    pickle_load, pickle_dump, time_consistent, update_time_db,
+    delete_cache_files, config_file_changed, update_config_file_db)
+
+
+class FileCache:
+    """
+    Example/Tutorial:
+
+    >>> from pyprint.NullPrinter import NullPrinter
+    >>> from coalib.output.printers.LogPrinter import LogPrinter
+    >>> import copy, time
+    >>> log_printer = LogPrinter(NullPrinter())
+
+    To initialize the cache create an instance for the project:
+
+    >>> cache = FileCache(log_printer, "test", flush_cache=True)
+
+    Now we can track new files by running:
+
+    >>> cache.track_new_files(["a.c", "b.c"])
+
+    Since all cache operations are lazy (for performance), we need to
+    explicitly write the cache to disk for persistence in future uses:
+    (Note: The cache will automatically figure out the write location)
+
+    >>> cache.write()
+
+    Let's simulate going into the future by doing ``sleep`` for a second:
+
+    >>> time.sleep(1)
+
+    Let's create a new instance to simulate a separate run:
+
+    >>> cache = FileCache(log_printer, "test", flush_cache=False)
+
+    >>> old_data = copy.deepcopy(cache.data)
+
+    We can mark a file as changed by doing:
+
+    >>> cache.add_to_changed_files({"a.c"})
+
+    Again write to disk after calculating the new cache times for each file:
+
+    >>> cache.write()
+
```

Comment:
Since the next line is unrelated to the tutorial text, I thought it would make sense to have a newline IMO.